h/accounts/views.py

```diff
@@ -132,7 +142,23 @@ def _logout(self):
 class AjaxAuthController(AuthController):
     def login(self):
         try:
-            appstruct = self.form.validate(self.request.json_body.items())
+            json_body = self.request.json_body
+        except ValueError as err:
+            raise accounts.JSONError(
+                _('Could not parse request body as JSON: {message}'.format(
+                    message=err.message)))
+
+        if not isinstance(json_body, dict):
+            raise accounts.JSONError(
+                _('Request JSON body must have a top-level object'))
+
+        # Transform non-string usernames and password into strings.
+        # Deform crashes otherwise.
+        json_body['username'] = unicode(json_body.get('username') or '')
+        json_body['password'] = unicode(json_body.get('password') or '')
```

Comment:
These lines are here because Deform crashes if username or body is not a string. Not sure if there's something that can be done to the schema to fix this:

```
_____________________________________________ test_login_ajax_returns_status_failure_on_non_string_password _____________________________________________

_ = <MagicMock name='check_csrf_token' id='140573433417360'>

    @pytest.mark.usefixtures('routes_mapper')
    @mock.patch('h.accounts.schemas.check_csrf_token')
    def test_login_ajax_returns_status_failure_on_non_string_password(_):
        for password in (None, 23, True):
            request = DummyRequest(json_body={
                'username': 'user', 'password': password})
    
            controller = AjaxAuthController(request)
    
>           result = controller.login()

h/accounts/test/views_test.py:292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
h/accounts/views.py:159: in login
    appstruct = self.form.validate(json_body.items())
../../../.virtualenvs/h/local/lib/python2.7/site-packages/deform-0.9.9-py2.7.egg/deform/field.py:602: in validate
    return self.validate_pstruct(pstruct)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/deform-0.9.9-py2.7.egg/deform/field.py:627: in validate_pstruct
    cstruct = self.deserialize(pstruct)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/deform-0.9.9-py2.7.egg/deform/field.py:478: in deserialize
    return self.widget.deserialize(self, pstruct)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/deform-0.9.9-py2.7.egg/deform/widget.py:1216: in deserialize
    result[name] = subfield.deserialize(subval)
../../../.virtualenvs/h/local/lib/python2.7/site-packages/deform-0.9.9-py2.7.egg/deform/field.py:478: in deserialize
    return self.widget.deserialize(self, pstruct)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <deform.widget.PasswordWidget object at 0x7fd9ce41e150>, field = <deform.field.Field object at 140573432840400 (schemanode 'password')>
pstruct = None

    def deserialize(self, field, pstruct):
        if pstruct is null:
            return null
        if self.strip:
>           pstruct = pstruct.strip()
E           AttributeError: 'NoneType' object has no attribute 'strip'

../../../.virtualenvs/h/local/lib/python2.7/site-packages/deform-0.9.9-py2.7.egg/deform/widget.py:267: AttributeError
```